spec/beaker/dsl/helpers/puppet_helpers_spec.rb

```diff
@@ -894,5 +894,92 @@ def stub_post_setup
+      subject.bounce_service( host, 'not_real_service', nil, 8000)
+    end
+  end
+
+  describe '#sleep_until_puppetdb_started' do
+    let( :options ) { Beaker::Options::Presets.new.presets }
+    before :each do
+      allow( subject ).to receive( :options ) { options }
+    end
+
+    it 'uses the default ports if none given' do
+      host = hosts[0]
+      expect( subject ).to receive( :curl_with_retries ).with( anything(), anything(), /8080/, anything(), anything() ).once.ordered
+      expect( subject ).to receive( :curl_with_retries ).with( anything(), anything(), /8081/, anything() ).once.ordered
+      subject.sleep_until_puppetdb_started( host )
+    end
+
+    it 'allows setting the nonssl_port' do
+      host = hosts[0]
+      expect( subject ).to receive( :curl_with_retries ).with( anything(), anything(), /8084/, anything(), anything() ).once.ordered
+      expect( subject ).to receive( :curl_with_retries ).with( anything(), anything(), /8081/, anything() ).once.ordered
+      subject.sleep_until_puppetdb_started( host, 8084 )
+
+    end
+
+    it 'allows setting the ssl_port' do
+      host = hosts[0]
+      expect( subject ).to receive( :curl_with_retries ).with( anything(), anything(), /8080/, anything(), anything() ).once.ordered
+      expect( subject ).to receive( :curl_with_retries ).with( anything(), anything(), /8085/, anything() ).once.ordered
+      subject.sleep_until_puppetdb_started( host, nil, 8085 )
+    end
+  end
+
+  describe '#sleep_until_puppetserver_started' do
+    let( :options ) { Beaker::Options::Presets.new.presets }
+    before :each do
+      allow( subject ).to receive( :options ) { options }
+    end
+
+    it 'uses the default port if none given' do
+      host = hosts[0]
+      expect( subject ).to receive( :curl_with_retries ).with( anything(), anything(), /8140/, anything() ).once.ordered
+      subject.sleep_until_puppetserver_started( host )
+    end
+
+    it 'allows setting the port' do
+      host = hosts[0]
+      expect( subject ).to receive( :curl_with_retries ).with( anything(), anything(), /8147/, anything() ).once.ordered
+      subject.sleep_until_puppetserver_started( host, 8147 )
+    end
+  end
+
+  describe '#sleep_until_nc_started' do
+    let( :options ) { Beaker::Options::Presets.new.presets }
+    before :each do
+      allow( subject ).to receive( :options ) { options }
+    end
+
+    it 'uses the default port if none given' do
+      host = hosts[0]
+      expect( subject ).to receive( :curl_with_retries ).with( anything(), anything(), /4433/, anything() ).once.ordered
+      subject.sleep_until_nc_started( host )
+    end
+
+    it 'allows setting the port' do
+      host = hosts[0]
+      expect( subject ).to receive( :curl_with_retries ).with( anything(), anything(), /4435/, anything() ).once.ordered
+      subject.sleep_until_nc_started( host, 4435 )
+    end
+  end
```

Comment:
There are definitely ways to write non-mockist spec tests; though the ability to do it varies with the specific code under test.  Sometimes working with existing code (as opposed to writing new code which could conceivably be done test-first or TDD with the tests influencing the design) can be difficult to test without resorting to stubs/mocks heavily. There's at least [one really good book](http://www.amazon.com/Working-Effectively-Legacy-Michael-Feathers/dp/0131177052) that I recommend regularly on the topic.

Sometimes difficulty of spec (or unit) testing code without mocks/stubs is a sign that refactoring to service classes, etc., might be helpful to find seams in the design where tests would work.

If I were to do work here to make the design more testable I would probably end up introducing at least one additional class -- either to act as a carrier for the various parameters, or a class to act as the final delegate to execute the curl functionality (independent of how it was called, etc.). 

Ironically `Command` classes tend to come into play for making code like our helpers more testable. I'm not sure that's the history behind the `Command` classes in Beaker, though.